contracts/src/CrossChainApplications/NativeTokenBridge/NativeTokenDestination.sol

```diff
@@ -0,0 +1,192 @@
+// (c) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+// SPDX-License-Identifier: Ecosystem
+
+pragma solidity 0.8.18;
+
+import "@openzeppelin/contracts/security/ReentrancyGuard.sol";
+import "@subnet-evm-contracts/interfaces/IWarpMessenger.sol";
+import "@subnet-evm-contracts/interfaces/IAllowList.sol";
+import "@subnet-evm-contracts/interfaces/INativeMinter.sol";
+import "./INativeTokenDestination.sol";
+import "../../Teleporter/ITeleporterMessenger.sol";
+import "../../Teleporter/ITeleporterReceiver.sol";
+import "../../Teleporter/SafeERC20TransferFrom.sol";
+
+// Native Minter Precompile Contract Address
+address constant MINTER_ADDRESS = 0x0200000000000000000000000000000000000001;
+// This is the address where the burned base fee go. We use this to send the amount of
+// burned tokens back to the source chain.
+// TODO implement mechanism to report burned tx fees to source chian.
+address constant BURNED_TX_FEES_ADDRESS = 0x0100000000000000000000000000000000000000;
+// Designated Blackhole Address. This is where we burn token before sending an unlock
+// message to the source chain. Different from the burned tx fee address so we can
+// track each separately.
+address constant BLACKHOLE_ADDRESS = 0x0100000000000000000000000000000000000001;
+// Precompiled Warp address
+address constant WARP_PRECOMPILE_ADDRESS = 0x0200000000000000000000000000000000000005;
+
+contract NativeTokenDestination is
+    ITeleporterReceiver,
+    INativeTokenDestination,
+    ReentrancyGuard
+{
+    using SafeERC20 for IERC20;
+
+    INativeMinter private immutable _nativeMinter =
+        INativeMinter(MINTER_ADDRESS);
+
+    uint256 public constant TRANSFER_NATIVE_TOKENS_REQUIRED_GAS = 150_000; // TODO this is a placeholder
+    bytes32 public immutable currentBlockchainID;
+    bytes32 public immutable sourceBlockchainID;
+    address public immutable nativeTokenSourceAddress;
+    // We will not mint the first `tokenReserve` tokens sent to this subnet.
+    // This should be constructed to match the initial token supply of this subnet.
+    // This will mean we will not mint tokens until the source contact is collateralized.
+    uint256 public tokenReserve;
```

Comment:
For naming clarity, I think `initialReserveImbalance` may better here (still trying to think of other options though).

"Token reserve" makes me think there is some amount of extra tokens reserved, but in this case we're referring to the opposite.

Reply:
Great! I knew this name wasn't great but couldn't come up with anything better. I like this one.

Reply:
Do you think we should have two different variables? `initialReserveImbalance` as a constant that is passed to the constructor, and `reserveImbalance` which is initialized to `initialReserveImbalance` but is decremented to 0 as we collateralize?

Reply:
Yeah I think that would make sense to allow for people to check what the initial pre-allocated supply was. Maybe:

```
uint256 public immutable preallocatedSupply;
uint256 public currentReserveImbalance;
```